### PR TITLE
Support dfont font files

### LINKED ISSUE #128: Support dfont
It's quite straightforward.  We should...

### PR DESCRIPTION
**Pull request by ebraminio**
Fixes #128. Hi @brawer @anthrotype I've tried to use your implementations on [GNU Classpath](https://github.com/jatovm/classpath/blob/master/gnu/java/awt/font/opentype/MacResourceFork.java) and [fonttools](https://github.com/fonttools/fonttools/pull/400/files) for this (which I derived [this](https://gist.github.com/ebraminio/3b1c91ce605cc93dba59d984ecece7b7) for convenience). [This](http://mirror.informatimago.com/next/developer.apple.com/documentation/mac/MoreToolbox/MoreToolbox-99.html) doc (which its original link was dead) was very helpful

### CODE REVIEW

File: src/hb-open-file-private.hh [hunk 1]
```diff
@@ -299,6 +375,7 @@ struct OpenTypeFontFile
     CFFTag		= HB_TAG ('O','T','T','O'), /* OpenType with Postscript outlines */
     TrueTypeTag	= HB_TAG ( 0 , 1 , 0 , 0 ), /* OpenType with TrueType outlines */
     TTCTag		= HB_TAG ('t','t','c','f'), /* TrueType Collection */
+    DFontTag		= HB_TAG ( 0 , 0 , 1 , 0 ), /* Mac Resource Fork */
```
Comment (ebraminio) [inline src/hb-open-file-private.hh]: It this a fact? It is true about the three remaining macOS dfont files, `/System/Library/Fonts/Courier.dfont`, `/System/Library/Fonts/Geneva.dfont` and ` /System/Library/Fonts/Monaco.dfont` but it seems is not that correct.

File: src/hb-open-file-private.hh [hunk 2]
```diff
@@ -286,6 +286,82 @@ struct TTCHeader
...
+  ArrayOf<ResourceTypeItem> types;
+  public:
+  DEFINE_SIZE_ARRAY (8, types);
+};
+
+struct ResourceForkHeader
+{
+  inline unsigned int get_face_count (void) const
+  {
+    return (((char*) this)+sizeof (ResourceForkHeader)+map).types.len + 1;
```
Comment (ebraminio) [inline src/hb-open-file-private.hh]: Can you speculatively guess what should be here? This doesn't feel right...

File: src/hb-open-file-private.hh [hunk 3]
```diff
@@ -286,6 +286,82 @@ struct TTCHeader
...
+ * Mac Resource Fork
+ */
+
+struct ResourceRefItem
+{
+  HBINT16 id; // Why is not unsigned?
+  HBINT16 nameOffset; // Why is not unsigned?
+  HBUINT8 attr;
+  char dataOffset[3];
+  HBUINT16 reserved;
```
Comment (ebraminio) [inline src/hb-open-file-private.hh]: Structs are copied from fonttools but I am not sure why signed number were there.

File: src/hb-open-file-private.hh [hunk 4]
```diff
@@ -286,6 +286,82 @@ struct TTCHeader
...
+struct ResourceForkHeader
+{
+  inline unsigned int get_face_count (void) const
+  {
+    return (((char*) this)+sizeof (ResourceForkHeader)+map).types.len + 1;
+  }
+
+  inline const OpenTypeFontFace& get_face (unsigned int i) const
+  {
+    return *new OpenTypeFontFace();//this+table[i];
```
Comment (ebraminio) [inline src/hb-open-file-private.hh]: Just a placeholder...

File: src/hb-open-type-private.hh [hunk 5]
```diff
@@ -672,7 +672,7 @@ typedef IntType<uint16_t, 2> HBUINT16;	/* 16-bit unsigned integer. */
 typedef IntType<int16_t,  2> HBINT16;	/* 16-bit signed integer. */
 typedef IntType<uint32_t, 4> HBUINT32;	/* 32-bit unsigned integer. */
 typedef IntType<int32_t,  4> HBINT32;	/* 32-bit signed integer. */
-typedef IntType<uint32_t, 3> UINT24;	/* 24-bit unsigned integer. */
+typedef IntType<uint32_t, 3> HBUINT24;	/* 24-bit unsigned integer. */
```
Comment (ebraminio) [inline src/hb-open-type-private.hh]: In order to harmonizing them and make it easier to find

**Comment by anthrotype** [comment]
I wrote that more than 2 years ago and I forgot most of it, sorry :)
You know this book right?
https://developer.apple.com/legacy/library/documentation/mac/pdf/MoreMacintoshToolbox.pdf

**Comment by ebraminio** [comment]
Oh interesting. Thank you! :)

File: src/hb-open-file-private.hh (hunk 1 above)
Comment (ebraminio) [reply src/hb-open-file-private.hh]: So maybe we should think about providing a separate API for this if so.

**Comment by ebraminio** [comment]
Happy to announce this is almost done :) (cc @simoncozens) just some testing and completing its sanitization is needed

**Comment by ebraminio** [comment]
Sanitization is done and made main.cc to run it on the `.dfont`s.

Currently `Courier.dfont` and `Monaco.dfont` are working great but `Geneva.dfont` don't, I should see why.

File: src/main.cc [hunk 6]
```diff
@@ -51,23 +51,33 @@ main (int argc, char **argv)
...
+  mm = HB_MEMORY_MODE_WRITABLE;
 #endif
 
   printf ("Opened font file %s: %d bytes long\n", argv[1], len);
 
-  const OpenTypeFontFile &ot = *CastP<OpenTypeFontFile> (font_data);
+  Sanitizer<OpenTypeFontFile> sanitizer;
+  hb_blob_t *blob = hb_blob_create (font_data, len, mm, (void *) font_data, destroy);
+  hb_blob_t *font_blob = sanitizer.sanitize (blob);
+  const OpenTypeFontFile &ot = *Sanitizer<OpenTypeFontFile>::lock_instance (font_blob);
```
Comment (ebraminio) [inline src/main.cc]: To test the sanitizer somewhere.

**Comment by anthrotype** [comment]
(out of curiousity, can all these dfont of yours be successfully decompiled with fonttools?)

**Comment by ebraminio** [comment]
> (out of curiousity, can all these dfont of yours be successfully decompiled with fonttools?)

fonttools dfont decoding was very helpful for this but the reason Geneva was failing was [this](https://github.com/harfbuzz/harfbuzz/blob/a0dccb6188d7a9174643c7239041cb1a5300b957/src/hb-open-type-private.hh#L820) that's why I replaced that part with `StructAtOffset` which wasn't an issue on fonttools at all :)

File: src/main.cc [hunk 7]
```diff
@@ -94,13 +107,13 @@ main (int argc, char **argv)
   printf ("%d font(s) found in file\n", num_fonts);
   for (int n_font = 0; n_font < num_fonts; n_font++) {
     const OpenTypeFontFace &font = ot.get_face (n_font);
-    printf ("Font %d of %d:\n", n_font, num_fonts);
+    printf ("Font %d of %d:\n", n_font + 1, num_fonts);
```
Comment (ebraminio) [inline src/main.cc]: I guess this looks better?

Comment (ebraminio) [reply src/main.cc]: Done

File: src/hb-open-file-private.hh [hunk 8]
```diff
@@ -286,6 +286,205 @@ struct TTCHeader
...
+  }
+
+  inline const ArrayOf<HBUINT8, HBUINT8>& get_name (const ResourceRefItem &item,
+						    unsigned int i) const
+  {
+    if (item.get_name_offset () == -1)
+      return Null<ArrayOf<HBUINT8, HBUINT8> >();
+
+    return (ArrayOf<HBUINT8, HBUINT8>&) ((this+nameList)[item.get_name_offset ()]);
+  }
```
Comment (ebraminio) [inline src/hb-open-file-private.hh]: `get_name` is not used now or ever but just to show the implementation is complete.

**Comment by ebraminio** [comment]
How about merging this Behdad but without hooking it up and deciding about that later?

**Comment by behdad** [comment]
We don't touch those parts of code often so it can wait a few more days of you don't mind.

File: src/hb-open-file-private.hh [hunk 9]
```diff
@@ -299,6 +501,7 @@ struct OpenTypeFontFile
     CFFTag		= HB_TAG ('O','T','T','O'), /* OpenType with Postscript outlines */
     TrueTypeTag	= HB_TAG ( 0 , 1 , 0 , 0 ), /* OpenType with TrueType outlines */
     TTCTag		= HB_TAG ('t','t','c','f'), /* TrueType Collection */
+    DFontTag		= HB_TAG ( 0 , 0 , 1 , 0 ), /* DFont Mac Resource Fork */
```
Comment (ebraminio) [inline src/hb-open-file-private.hh]: Obviously this is not standard but every .dfont file I could see have it. If we don't agree on it we should provide a different API for whole dfont thing.

File: src/hb-open-file-private.hh [hunk 10]
```diff
@@ -286,6 +286,208 @@ struct TTCHeader
...
+struct ResourceRefItem
+{
+  inline bool sanitize (hb_sanitize_context_t *c) const
+  {
+    TRACE_SANITIZE (this);
+    // actual data sanitization is done on ResourceForkHeader sanitizer
+    return_trace (c->check_struct (this));
+  }
+
+  inline unsigned int get_data_offset () const
```
[the remark this answers is not in the document]
Comment (ebraminio) [reply src/hb-open-file-private.hh]: Done

File: src/hb-open-file-private.hh [hunk 11]
```diff
@@ -286,6 +286,208 @@ struct TTCHeader
...
+struct ResourceTypeItem
+{
+  inline bool sanitize (hb_sanitize_context_t *c) const
+  {
+    TRACE_SANITIZE (this);
+    // RefList sanitization is done on ResourceMap sanitizer
+    return_trace (c->check_struct (this));
+  }
+
+  inline unsigned int get_number_of_resource () const
```
[the remark this answers is not in the document]
Comment (ebraminio) [reply src/hb-open-file-private.hh]: `get_resource_count`, done

File: src/hb-open-file-private.hh [hunk 12]
```diff
@@ -286,6 +286,208 @@ struct TTCHeader
...
+  }
+
+  inline const ResourceRefItem& get_ref_item (const void *base,
+					      unsigned int i) const
+  {
+    return (base+refList)[i];
+  }
+
+  protected:
+  HBUINT8	type[4];	/* Resource type */
```
[the remark this answers is not in the document]
Comment (ebraminio) [reply src/hb-open-file-private.hh]: Done.

File: src/hb-open-file-private.hh [hunk 13]
```diff
@@ -286,6 +286,208 @@ struct TTCHeader
...
+    return_trace (true);
+  }
+
+  inline const ResourceTypeItem& get_type (unsigned int i) const
+  {
+    // Why offset from the second byte of the object? I'm not sure
+    return ((&reserved[2])+typeList)[i];
+  }
+
+  inline unsigned int get_number_of_types () const
```
[the remark this answers is not in the document]
Comment (ebraminio) [reply src/hb-open-file-private.hh]: Done

File: src/hb-open-file-private.hh [hunk 14]
```diff
@@ -286,6 +286,208 @@ struct TTCHeader
...
+    return nTypes + 1;
+  }
+
+  inline const ResourceRefItem &get_ref_item (const ResourceTypeItem &type,
+					      unsigned int i) const
+  {
+    return type.get_ref_item (&(this+typeList), i);
+  }
+
+  inline const ArrayOf<HBUINT8, HBUINT8>& get_name (const ResourceRefItem &item,
```
[the remark this answers is not in the document]
Comment (ebraminio) [reply src/hb-open-file-private.hh]: Done, couldn't find the `post` table use for a quick look however.

File: src/main.cc (hunk 6 above)
Comment (ebraminio) [reply src/main.cc]: Done. #994

**Comment by ebraminio** [comment]
PTAL

**Comment by ebraminio** [comment]
Rebased

**Pull request merged**